Compiler/src/typeinfer.jl

```diff
@@ -145,9 +145,13 @@ function finish!(interp::AbstractInterpreter, caller::InferenceState;
         if !@isdefined di
             di = DebugInfo(result.linfo)
         end
-        ccall(:jl_update_codeinst, Cvoid, (Any, Any, Int32, UInt, UInt, UInt32, Any, UInt8, Any, Any),
-            ci, inferred_result, const_flag, first(result.valid_worlds), last(result.valid_worlds), encode_effects(result.ipo_effects),
-            result.analysis_results, relocatability, di, edges)
+        min_world, max_world = first(result.valid_worlds), last(result.valid_worlds)
+        ipo_effects = encode_effects(result.ipo_effects)
+        ccall(:jl_update_codeinst, Cvoid, (
+            Any, Any, Any, Any, Any, Int32, UInt, UInt,
+            UInt32, Any, UInt8, Any, Any),
+            ci, inferred_result, rettype, exctype, rettype_const, const_flags, min_world, max_world,
```

Comment:
Changing any non-atomic fields here (after `cache_result`) causes codegen to become unsound. I am uncertain how we would preserve the expectation that the optimizer only uses CodeInstances found in the cache, while also delaying publishing them into the cache until after this point in the code. I don't believe it to be impossible to change that (since we are holding the "engine lock" on this until after the optimizer finishes), just tricky to figure out how to correct adapt to this PR.